java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,157 @@
+    if (!hasSemantic() || !(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    ClassTree classTree = (ClassTree) tree;
+    String classPackageName = packageNameOf(classTree);
+
+    Optional<AnnotationTree> componentScanAnnotation = classTree.modifiers().annotations().stream()
+      .filter(SpringBeansShouldBeAccessibleCheck::isComponentScan).findFirst();
+    if (componentScanAnnotation.isPresent()) {
+      componentScanAnnotation.get().arguments().forEach(this::addToScannedPackages);
+    } else if (hasAnnotation(classTree, SPRING_BOOT_APP_ANNOTATION)) {
+      scannedPackages.add(classPackageName);
+    } else if (hasAnnotation(classTree, SPRING_BEAN_ANNOTATIONS)) {
+      addMessageToMap(classPackageName, classTree);
+    }
+  }
+
+  private void addMessageToMap(String classPackageName, ClassTree classTree) {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    IdentifierTree className = classTree.simpleName();
+    if (className != null) {
+      AnalyzerMessage analyzerMessage = reporter.createAnalyzerMessage(this, className, String.format(MESSAGE_FORMAT, className));
+      messagesPerPackage.computeIfAbsent(classPackageName, k -> new ArrayList<>());
+      messagesPerPackage.get(classPackageName).add(analyzerMessage);
+    }
+  }
+
+  private void addToScannedPackages(ExpressionTree annotationArgument) {
+    if (annotationArgument.is(Tree.Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree argumentAssignment = (AssignmentExpressionTree) annotationArgument;
+      ExpressionTree argumentAssignmentVar = argumentAssignment.variable();
+      if (argumentAssignmentVar.is(Tree.Kind.IDENTIFIER) && COMPONENT_SCAN_ARGUMENTS.contains(((IdentifierTree) argumentAssignmentVar).name())) {
+        addLiteralsToScannedPackages(argumentAssignment.expression());
+      }
+    } else {
+      addLiteralsToScannedPackages(annotationArgument);
+    }
+  }
+
+  private void addLiteralsToScannedPackages(ExpressionTree packageNames) {
+    if (packageNames.is(Tree.Kind.STRING_LITERAL)) {
+      String name = ConstantUtils.resolveAsStringConstant(packageNames);
+      scannedPackages.add(name);
+    } else if (packageNames.is(Tree.Kind.NEW_ARRAY)) {
+      for (ExpressionTree p : ((NewArrayTree) packageNames).initializers()) {
+        String name = ConstantUtils.resolveAsStringConstant(p);
+        scannedPackages.add(name);
+      }
+    }
+  }
+
+  private static String packageNameOf(ClassTree classTree) {
+    Tree classTreeParent = classTree.parent();
+    if (classTreeParent != null && classTreeParent.is(Tree.Kind.COMPILATION_UNIT)) {
+      return PackageUtils.packageName(((CompilationUnitTree) classTreeParent).packageDeclaration(), ".");
+    }
+    return "";
+  }
+
+  private static boolean hasAnnotation(ClassTree classTree, String... annotationName) {
+    return Arrays.stream(annotationName).anyMatch(annotation -> classTree.symbol().metadata().isAnnotatedWith(annotation));
```

Comment:
extract `classTree.symbol().metadata()` to not recompute it for each annotation.

Reply:
good catch!